filesystem/filesystem.go

```diff
@@ -129,6 +133,23 @@ func (m *Mount) BaseDir() string {
 	return filepath.Join(m.Path, baseDirName)
 }
 
+// baseDirReal is like BaseDir, except it evaluates the symlink if the BaseDir
+// is a symlink.
+func (m *Mount) baseDirReal() (string, error) {
```

Comment:
So this approch will work, but it seems like it would be simpler to just have `BaseDir` call [`filepath.EvalSymlinks`](https://golang.org/pkg/path/filepath/#EvalSymlinks) and just return that for all `BaseDir` operations. That would also help keep the symlink code size down and in one place.

```go
// BaseDir returns the path of the base fscrypt directory on this filesystem after resolving symlinks.
func (m *Mount) BaseDir() string {
	dir, err := filepath.EvalSymlinks(filepath.Join(m.Path, baseDirName))
        if err != nil {
                panic(err)
        }
        return dir
}

Reply:
That doesn't work because `EvalSymlinks` returns an error if the symlink target doesn't exist.  So I went with:
```go
func (m *Mount) BaseDir() string {
        rawBaseDir := filepath.Join(m.Path, baseDirName)
        target, err := os.Readlink(rawBaseDir)
        if err != nil {
                return rawBaseDir // not a symlink
        }
        if filepath.IsAbs(target) {
                return target
        }
        return filepath.Join(m.Path, target)
}
```

One thing I don't quite like about either approach is that it causes `BaseDir()` (which is called several times on each mount) to do syscalls rather than just return a string. The single `readlink()` isn't too bad in comparison to the full path walk done by `EvalSymlinks()` though, so I suppose that for simplicity in the code we should just go with that for now.

Reply:
This is much easier for me to understand. If it turns out `BaseDir()` making syscalls is a bottleneck for the `fscrypt` commands, we can always just cache the result (so that `readLink` is only ever called once per mount).